### ISSUE TITLE
[Potential] error in computing attention and final word probability

### ISSUE THREAD

**Issue by hashbangCoder** [abisee/pointer-generator#41]
Hi, 

In the *attention_decoder.py*, you compute attention scores over the encoder states from [L92](https://github.com/abisee/pointer-generator/blob/master/attention_decoder.py#L92) - [L103](https://github.com/abisee/pointer-generator/blob/master/attention_decoder.py#L103). 

I can't seem to find where you've masked the encoder features of size - (batch_size,attn_length, attention_vec_size). Without masking, the computed attention score will be incorrect. It will assign values to encoder states that correspond to PAD tokens upon taking softmax. 

Also, the word probability over the extended vocabulary is computed using `tf.scatter_nd`. This does not work in case of duplicate indices as the output will be non-deterministic and not summed up.. So, if the source text had more than occurence of an OOV word then it will become a problem.

**Comment by abisee** [comment]
Hi @hashbangCoder, thanks for the questions.

**About the encoder padding:** yes I think you're right. As the code stands, each encoder hidden state (including those corresponding to PAD tokens) receives an attention score, and the softmax  calculation on [this line](https://github.com/abisee/pointer-generator/blob/master/attention_decoder.py#L103) will include those PAD locations. So the attention distribution sums to 1. 

To fix this, I think it would be as simple as just inserting a line [here](https://github.com/abisee/pointer-generator/blob/master/attention_decoder.py#L101) and [here](https://github.com/abisee/pointer-generator/blob/master/attention_decoder.py#L110) that applies a mask to `e`, setting `e` to zero in the PAD locations.

I just ran a quick experiment to see to what degree this has an effect when you run the code with the default settings and data. Firstly, all articles are truncated to first 400 tokens, but almost all articles have more than 400 tokens, so they have no padding. This means that for almost all examples, this error does not make a difference. Second, I observe that for those examples that are shorter than 400 tokens, the attention on the PAD locations starts small and then quickly decreases to zero. So I don't think this error has a large effect when you run this particular experiment.

**About scatter_nd**: I don't understand your question. Why does it not work if there are duplicate instances of an OOV word?

**Comment by hashbangCoder** [comment]
Thanks for the reply! 
I have a solution for the attention bug. You could just mask with (-inf) before applying softmax. This will ensure the padded states are assigned 0 probability. Don't know you'd do it in TF, but am sure its possible.

Well, I'm using Pytorch and the `scatter`in that doesn't work with duplicates and overwrites to the last instance of the index. I looked up Tensorflow's scatter and while it made no mention of this, I wasn't sure if it indeed sums all instances of the index. 
Sorry about that.

**Comment by abisee** [comment]
Yes, I just realized that setting to zero before softmax is not correct. You either need to set to -inf before softmax (that feels pretty hacky though) or set to zero after softmax, then re-normalize. I've implemented the latter and am testing it out.

`tf.scatter_nd` performs addition; see [this discussion](https://github.com/abisee/pointer-generator/issues/22).

**Comment by hashbangCoder** [comment]
I would disagree about `-inf` feeling hacky for example openNMT in their Pytorch implementation use it for masking. Im curious as to how you're renormalizing after SoftMax i.e how you're redistributing the probability of masked out values to the remaining elements. 

Also, I've implemented your model in Pytorch and am training it currently. I'm surprised by how you managed to fir the entire model (50,000 `vocab_size`, 256 `hidden_size`, 140 `max_abstract` and 400 `max_article`) on one GPU. 
I had to reduce my vocabulary to 25,000 to accommodate the other parameters on an Nvidia TitanX with 12G of memory.  Maybe it has to do with the fact that you're using Adagrad which might be using less memory than Adam (not sure of this since both track param history)

**Comment by abisee** [comment]
@hashbangCoder The matter of masking the encoder states before taking attention has been fixed in the [latest commit](https://github.com/abisee/pointer-generator/commit/d08c4c5cc358a0e9bdeebb46e47885cd8cdb2760).

**Comment by hashbangCoder** [comment]
@abisee 
Thanks for the bug fix! It would be helpful if you could answer the memory intensive nature of the model above.

**Comment by abisee** [comment]
@hashbangCoder I also had problems fitting everything in GPU memory, and as far as I can remember the most important hyperparameters were `max_enc_steps` and `vocab_size`. Is your word embedding size the same too? Perhaps the GPU we were using has a greater memory capacity than the one you're using? Unfortunately I am not an expert in understanding how GPU memory works, but I believe it's often be more complicated than you might think. 

Good luck!
